src/node_http2.cc

```diff
@@ -1292,45 +1292,30 @@ void Http2Session::HandleHeadersFrame(const nghttp2_frame* frame) {
   Local<String> value_str;
 
   Local<Array> holder = Array::New(isolate);
-  Local<Function> fn = env()->push_values_to_array_function();
-  Local<Value> argv[NODE_PUSH_VAL_TO_ARRAY_MAX * 2];
-
   // The headers are passed in above as a queue of nghttp2_header structs.
   // The following converts that into a JS array with the structure:
   // [name1, value1, name2, value2, name3, value3, name3, value4] and so on.
   // That array is passed up to the JS layer and converted into an Object form
   // like {name1: value1, name2: value2, name3: [value3, value4]}. We do it
   // this way for performance reasons (it's faster to generate and pass an
   // array than it is to generate and pass the object).
-  size_t n = 0;
-  while (n < headers.size()) {
-    size_t j = 0;
-    while (n < headers.size() && j < arraysize(argv) / 2) {
-      nghttp2_header item = headers[n++];
-      // The header name and value are passed as external one-byte strings
-      name_str =
-          ExternalHeader::New<true>(this, item.name).ToLocalChecked();
-      value_str =
-          ExternalHeader::New<false>(this, item.value).ToLocalChecked();
-      argv[j * 2] = name_str;
-      argv[j * 2 + 1] = value_str;
-      j++;
-    }
-    // For performance, we pass name and value pairs to array.protototype.push
-    // in batches of size NODE_PUSH_VAL_TO_ARRAY_MAX * 2 until there are no
-    // more items to push.
-    if (j > 0) {
-      fn->Call(env()->context(), holder, j * 2, argv).ToLocalChecked();
-    }
+  size_t headers_size = headers.size();
+  std::vector<Local<Value>> headers_v(headers_size * 2);
+  for (size_t i = 0; i < headers_size; ++i) {
```

Comment:
In general I prefer pre-allocating, as range-for merely offer any benefit in these cases (I don't think it's any harder to read than `push_back` and it's even a bit easier to do the math mentally with the `i * 2` thing) while pre-allocating at least gives a little performance boost.

Reply:
[ES.71](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-for-range): Prefer a range-for-statement to a for-statement when there is a choice
[ES.55](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-range-checking): Avoid the need for range checking
(we chose to adopt the C++CG to avoid personal preferences and bikeshedding ¯\_(ツ)_/¯)

As for `i`, it's not needed, so IMHO not introducing an unnecessary variable, is more readable.

Reply:
@refack This is not just a style issue as there is actual difference in how the code allocates memory, in my opinion at least this kind of obvious efficiency should have a higher precedence over stylish preferences.

Reply:
> Look at loops, if a traditional loop just looks at each element of a sequence, and there are no side effects on what it does with the elements, rewrite the loop to a ranged-for loop.

Also, the C++CG does not apply in this case since the code in the loop results in side effects (vector growth/allocation)

Reply:
> in my opinion at least this kind of obvious efficiency should have a higher precedence over stylish preferences.

Since this is a hot path, I agree that performance should be considered. But the C++CG are not just style, they are motivated by safety, correctness as well.

As for this case, pre allocating doesn't conflict with safety:
```cpp
std::vector<Local<Value>> headers_v();
headers_v.reserve(headers.size() * 2);
for (const nghttp2_header& item: headers) {
  headers_v.push_back(ExternalHeader::New<true>(this, item.name).ToLocalChecked());
  headers_v.push_back(ExternalHeader::New<false>(this, item.value).ToLocalChecked());
}
```
BTW `std::vector<Local<Value>> headers_v(headers_size * 2);` runs the `Local<Value>` ctor `headers_size * 2` time, while reserve doesn't ;)

Reply:
> BTW std::vector<Local<Value>> headers_v(headers_size * 2); runs the Local<Value> ctor headers_size * 2 time, while reserve doesn't ;)

`Local<>` is just a wrapper of pointers so I didn't really give it too much thought ¯\\_(ツ)_/¯ but yeah I guess why not

Reply:
Hmm, but using `push_back` will lose the benefit of self-documenting equations..

Reply:
@refack 

> I really like it that we have a document that lists what things do

To a certain degree, I agree, but without automatic tooling, this may just introduce more nits than necessary.